DEP: Deprecate 'generic' unit in np.timedelta64

riku-sakamoto opened the pull request:
This PR deprecates the `generic` unit in `np.timedelta64`. Using this unit can lead to unexpected behavior in some cases (see #28287 for details). Using `generic` unit now raises a `DeprecationWarning`.

* Allowed behavior (no warning):

```python
import numpy as np

np.timedelta64(10, 's')  # seconds
np.timedelta64("NaT", "ns") # NaT is also created with unit
```

* Deprecated behavior (raises FutureWarning):

```python
import numpy as np

np.timedelta64(10)  # generic unit
np.timedellta64("NaT") # NaT without generic unit is not allowed
np.timedelta64(10, 's') + 5 # adding integer to timedelta with generic unit
```

## Changes

### Main

- Raise `DeprecationWarning`  when `np.timedelta64` is constructed with the `generic` unit.
- Add tests to ensure the warning is raised in relevant scenarios.
- Add Release notes
- Improve not to raise the warning in `np.ones` and `np.ones_like` .

### Test

Test updates fall into three categories.

- Suppressing the warning in tests where the `generic` unit seems to be used intentionally. (`with pytest.warns(DeprecationWarning)`)

- Adding tests to check `generic` unit's behavior. Some tests uses `generic` timedelta in `pytest.mark.parametrize`. In such cases, the warnings cannot be suppressed with `pytest.mark.filterwarnings`. Instead, this PR add additional tests to check only `generic` unit's behavior.

- Modifying tests to use an explicit unit instead of `generic`.


## Future follow up items

- `np.timedelta64()` creates a `generic` timedelta with value `0`. We may want to change this to create a timedelta with an explicit unit (e.g., `np.timedelta64(0, 's')`) instead.

- ~~`np.ones_like` now raises `DeprecationWarning` when the input array is of timedelta type.~~ `Resolved`

- ~~`np.all_close` now raises `DeprecationWarning` when the input arrays are timedelta type.~~ `Resolved`

riku-sakamoto committed:
ENH: Raise RuntimeWarning when casting from a unitless `np.timedelta64` to unitful one.

This is a temporary solution to handle #28287.
Rasing `RuntimeWarning` encourages users to pay attention to the behavior.

tylerjereddy commented on pytest.ini:
```diff
@@ -27,6 +27,8 @@ filterwarnings =
 # Ignore DeprecationWarning from typing.mypy_plugin
     ignore:`numpy.typing.mypy_plugin` is deprecated:DeprecationWarning
+# Ignore Runtime Warning from datetime by calculating unitless value and unitful value
+    ignore:Casting from unitless timedelta to unitful timedelta is ambiguous.:RuntimeWarning
```
I believe that if we do this, we'd likely want at least one test case that actually makes sure that the warning is issued.

I see locally that there are 18 failures if this is removed though, and we wouldn't want 18 checks for the warning. Still, it might be nice if we could suppress most of them but enforce at least one of them.

That said, the discussion in the matching issue suggests that the decision here may be tricky, so it may be best to wait for some design feedback first before making changes.

riku-sakamoto replied:
According to the issue discussion and the triage meeting, deprecation of the np.generic unit has been decided. I have updated the test implementation as you suggested.

I would appreciate it if you could take a look.

riku-sakamoto committed:
REV: warning when calulating generic unit

riku-sakamoto committed:
DEP: deprecate the ``generic`` unit in `numpy.timedelta64` and `numpy.datetime64`

The ``generic`` unit in `numpy.timedelta64` and `numpy.datetime64` is now
deprecated. Using it will raise a ``FutureWarning``. (See #28287)

riku-sakamoto committed:
ENH: remove unnecessary `if` condition.

riku-sakamoto commented:
> Needs a release note.

I've added the release note. Thank you for the review!

riku-sakamoto committed:
TST: Improve tests for deprecation of generic unit

* Asserting warnings with the warns function¶
* Some tests are moved to `test_deprecations.py`

riku-sakamoto commented on numpy/_core/tests/test_cython.py:
```diff
@@ -83,6 +83,7 @@ def install_temp(tmpdir_factory):
     sys.path.append(str(build_dir))
 
 
+@pytest.mark.filterwarnings("ignore::FutureWarning")
```
Thanks for the comment! I've updated the test to assert that the warning is raised when expected.

`pytest.mark.filterwarnings` is now used only in `numpy/typing/tests/test_typing.py` to avoid failures when this test loads a Python script file that uses the `generic` unit. I can update it further if needed.

riku-sakamoto committed:
FIX: Tweak error emitting condition

riku-sakamoto committed:
DEP: Deprecate "NaT" created with a unitless form

riku-sakamoto commented:
This branch is far behind current main and it has conflicts.
So, I am going to rebase it.

riku-sakamoto committed:
TST: update tests

riku-sakamoto committed:
[ENH] improve not to emit warning when np.all_close is called for timedelta with unit

riku-sakamoto committed:
[ENH] improve not to raise warning when calling `np.ones` for timedelta with unit

riku-sakamoto committed:
DOC: improve release note

seberg committed:
Merge branch 'main' into enhancement/raise_warning_when_generic_unit_casting

seberg reviewed (approved):
Thanks, we had discussed this a few times and in general agreed to try this.

I'll ping the mailing list as well. If there is downstream fallout, we should consider reverting this.  From pandas perspective this is the right step, but I am not 100% sure if it a painless one either way.
Astropy might also notice it (but they test against nighties well).

In general, removing the ability to mix integers with timedelta/datetimes seems right and this is probably the most gentle first step.

We _could_ even re-instate specific things like `arr + 1` e.g. ufuncs and comparisons that mix with integers (or just Python integers).
